core/src/main/scala/kafka/coordinator/group/CoordinatorPartitionWriter.scala

```diff
@@ -66,6 +65,10 @@ class CoordinatorPartitionWriter[T](
   compressionType: CompressionType,
   time: Time
 ) extends PartitionWriter[T] {
+  private val threadLocalBufferSupplier = ThreadLocal.withInitial(
+    () => new BufferSupplier.GrowableBufferSupplier()
+  )
```

Comment:
some general questions for my understanding:

1. We are introducing a buffer supplier so that we can reuse the same buffer and not reallocate a new buffer each time.
2. It's a growable buffer supplier since maxBatchSize may change on subsequent append()
3. we're adding a try-finally block to always clear the buffer in the case of exceptions. this was not needed previously since we were not using a buffer supplier
4. it's a thread local supplier because we don't want the buffer to be reused across different threads - i don't think i fully understand the repercussions

is 1-3 correct? 4 is the actual question. thanks!

Reply:
For more context on 4. There are assumptions in the code that a request is handled on the same thread. 
There was a bug in KIP-890 (see [here](https://issues.apache.org/jira/browse/KAFKA-15653)) where we used the wrong byte buffer and it caused NPEs and other exceptions.

Reply:
makes sense. thanks!

Reply:
Just for my understanding though -- in the past we relied on RequestLocal for this sort of thing. Is the idea to move to ThreadLocal here so it is safer and we don't run into things like https://issues.apache.org/jira/browse/KAFKA-15653 ?

Reply:
That’s right.